lib/core/connector_job.rb

```diff
@@ -27,12 +28,38 @@ def self.pending_jobs(page_size = DEFAULT_PAGE_SIZE)
       fetch_jobs_by_query(query, page_size)
     end
 
-    def self.orphaned_jobs(_page_size = DEFAULT_PAGE_SIZE)
-      []
+    def self.orphaned_jobs(page_size = DEFAULT_PAGE_SIZE)
+      connector_ids = ConnectorSettings.fetch_native_connectors.map(&:id)
+      query = { bool: { must_not: { terms: { 'connector.id': connector_ids } } } }
+      fetch_jobs_by_query(query, page_size)
     end
 
-    def self.stuck_jobs(_page_size = DEFAULT_PAGE_SIZE)
-      []
+    def self.fetch_by_connector_id(connector_id, page_size = DEFAULT_PAGE_SIZE)
+      query = { term: { 'connector.id': connector_id } }
+      fetch_jobs_by_query(query, page_size)
+    end
+
+    def self.cleanup_jobs(jobs)
+      query = { terms: { '_id': jobs.map(&:id) } }
+      ElasticConnectorActions.delete_jobs_by_query(query)
+    end
+
+    def self.stuck_jobs(connector_id = nil, page_size = DEFAULT_PAGE_SIZE)
+      connector_ids = if connector_id
+                        [connector_id]
+                      else
+                        ConnectorSettings.fetch_native_connectors.map(&:id)
+                      end
+      query = {
+        bool: {
+          filter: [
+              { terms: { 'connector.id': connector_ids } },
+              { terms: { status: Connectors::SyncStatus::ACTIVE_STATUSES } },
+              { range: { last_seen: { lte: "now-#{STUCK_THRESHOLD}s" } } }
```

Comment:
>How frequently do we expect `last_seen` to be updated?

Every 10 seconds
https://github.com/elastic/connectors-ruby/blob/b3cc1332879a38930a272a63f8c6be1847578204/lib/core/sync_job_runner.rb#L24

>30 seconds doesn't feel like a very long time.
@danajuratoni suggested 1 minute [here](https://github.com/elastic/enterprise-search-team/issues/3199#issuecomment-1314068902), and I will make the change.

>Especially if two server clocks are not perfectly synced.

Good point.

Note, both should be configurable in `_meta` of  `.elastic-connectors`, but we deferred it post 8.6

Reply:
I'd put 5 minutes as a reasonable default TBH

Reply:
The heartbeat for a job is 10 seconds. If a job is stuck, it will take 5 mins to be identified. What should be the interval to run the clean-up task (the current default is 5 mins). 

With the above, a stuck job will only be cleaned worst case 10 mins. I will leave this to PM @danajuratoni :

How long we should wait to mark a job as stuck - m mins.
How often we should run the clean-up job? - n mins.

Ideally, a job will be marked as stuck after m mins without updating. In reality, it will be from m mins (best case) to (m+n) mins (worst case)

Reply:
As we don't have any data atm how often stuck jobs would occur and the impact of our checks, I tend to be more conservative with the configured times. The important thing is to adjust them when we have more insights.
From a product perspective, I'd suggest to use 1 minute for stuck jobs and 5 minutes for the cleanup job and make sure in 8.7 we have telemetry to capture related data.

Reply:
If you run MongoDB connector against our cloud instance sometimes new batch of docs take 3 minutes to get back. In reality I see some connectors hang for up to 5 minutes for some weird systems that throttle by keeping open connection and sending no data, thus I proposed 5 minutes.

Rrunning job clean up every hour or 24 hours might be good enough, but it's not a strong opinion.